File: scala/leet_code/1289.scala
/** This object represents a solution for finding the minimum falling path sum in a grid.
  * A falling path starts at any element in the first row and chooses the element in the next row which is either directly below or diagonally below to the left or right.
  * The goal is to reach the last row with the minimum sum of the chosen path.
  */
object Solution {

  /** Finds the minimum falling path sum in a grid.
    *
    * This method iterates over the grid, updating the sumRow array with the minimum sum of the falling path up to the current row.
    * It uses a helper function to find the indices of the minimum and the second minimum elements in the sumRow array.
    * It then updates the current row by adding the minimum sum of the falling path that ends at each element.
    *
    * @param grid The input grid. It should be a 2D array of integers.
    * @return The minimum falling path sum in the grid. It returns an integer.
    */
  def minFallingPathSum(grid: Array[Array[Int]]): Int = if (grid.length == 1)
    grid.head.head
  else {

    /** Finds the indices of the minimum and the second minimum elements in a row.
      *
      * This method iterates over the row, keeping track of the indices of the minimum and the second minimum elements.
      *
      * @param row The input row. It should be an array of integers.
      * @return A tuple of two integers representing the indices of the minimum and the second minimum elements in the row.
      */
    def findMinInds(row: Array[Int]): (Int, Int) =
      row.indices.foldLeft((0, 1)) {
        case ((minInd, nextMinInd), i) if row(i) < row(minInd) => (i, minInd)
        case ((minInd, nextMinInd), i)
            if row(i) < row(nextMinInd) && i != minInd =>
          (minInd, i)
        case (mins, _) => mins
      }

    grid.reduceLeft { (sumRow, row) =>
      val (minInd, nextMinInd) = findMinInds(sumRow)
      val min = sumRow(minInd)
      for (i <- row.indices)
        row(i) += (if (i != minInd) min else sumRow(nextMinInd))
      row
    }.min
  }
}

File: python/leet_code/1289.py
"""leet_code/1289.py

This module contains a solution for finding the minimum falling path sum in a given grid.
The grid is a 2D list of integers where each element represents a cost. A falling path starts from any element in the first row and chooses the element from the next row that is either directly below or diagonally left/right. Specifically, the path must be from one of the three cells in the row below. The same cell may not be chosen twice.

The solution uses dynamic programming to keep track of the minimum cost path at each cell.

Functions:
    minFallingPathSum: This function finds the minimum falling path sum in a grid.
"""

from json import loads
from sys import stdin
from typing import List


def minFallingPathSum(grid: List[List[int]]) -> int:
    """Finds the minimum falling path sum in a grid.

    Args:
        grid (List[List[int]]): The grid in which to find the minimum falling path sum.

    Returns:
        int: The minimum falling path sum in the grid.

    This function does not handle errors explicitly. If the input is not a 2D list of integers, Python's built-in error handling will raise an exception.
    """

    n = len(grid)
    next_min1_col = -1
    next_min2_col = -1
    next_min1_val = -1
    next_min2_val = -1

    # Find the minimum cost in the last row
    for col in range(n):
        if next_min1_col == -1 or grid[n - 1][col] <= next_min1_val:
            next_min2_col = next_min1_col
            next_min2_val = next_min1_val

            next_min1_col = col
            next_min1_val = grid[n - 1][col]

        elif next_min2_col == -1 or grid[n - 1][col] <= next_min2_val:
            next_min2_col = col
            next_min2_val = grid[n - 1][col]

    # Find the minimum cost in the remaining rows
    for row in range(n - 2, -1, -1):
        min1_col = -1
        min2_col = -1
        min1_val = -1
        min2_val = -1

        for col in range(n):
            ans = 0
            if col != next_min1_col:
                ans = grid[row][col] + next_min1_val
            else:
                ans = grid[row][col] + next_min2_val

            if min1_col == -1 or ans <= min1_val:
                min2_col = min1_col
                min2_val = min1_val

                min1_col = col
                min1_val = ans
            elif min2_col == -1 or ans <= min2_val:
                min2_col = col
                min2_val = ans

        next_min1_col = min1_col
        next_min1_val = min1_val
        next_min2_col = min2_col
        next_min2_val = min2_val

    return next_min1_val


f = open('user.out', 'w')
for i in map(loads, stdin):
    f.write(f'{minFallingPathSum(i)}\n')
f.flush()
exit(0)
